File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */


class Solution {
public:
    int GCD(int a, int b) {
        if (a == 1 || b == 1)
            return 1;
        while (b != 0){
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* temp = head;
        while (temp->next != NULL) {
            int gcd = GCD(temp->val, temp->next->val);
            ListNode* mid = new ListNode(gcd, temp->next);
            temp->next = mid;
            temp = mid->next;
        }

        return head;
    }
};

File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */


class Solution {
    public int GCD(int a, int b) {
        if (b == 0)
            return a;
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        ListNode temp = head;
        while (temp.next != null) {
            int gcd = GCD(temp.val, temp.next.val);
            ListNode mid = new ListNode(gcd, temp.next);
            temp.next = mid;
            temp = mid.next;
        }

        return head;
    }
}